### PR TITLE
feat: submissions overview mail

### PR DESCRIPTION
**Pull request by laurensmertens**
|Q              |A  |
|---------------|---|
|Bug fix?       |No|
|New feature?   |Yes|	
|BC breaks?     |No|	
|Deprecations?  |No|	
|Fixed tickets  |No|	

Added a service method to send a list of submissions by mail and a command to use it.

### CODE REVIEW

File: Command/EmailSubmissionsCommand.php [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        $formInstance = \strval($input->getArgument('formInstance'));
+        $templateId = \strval($input->getArgument('templateId'));
+        $emails = (array) $input->getArgument('emails');
+
+        $submissions = $this->formSubmissionService->getFormInstanceSubmissions($formInstance);
+
+        $this->formSubmissionService->mailSubmissions($submissions, $formInstance, $templateId, $emails);
```
Comment (dameert) [inline Command/EmailSubmissionsCommand.php:50]: I think that a formSubmissionService should fetch the data of all submissions. And a separate MailService should send the mail. That way sending mails becomes reusable without extra effort.

And we could imagine to export formSubmissions via other channels then mails.

File: Service/Form/Submission/FormSubmissionService.php [hunk 2]
```diff
@@ -91,4 +131,52 @@ public function submit(FormSubmissionRequest $submitRequest): array
...
+            return 'There are no submissions for this form';
+        }
+
+        try {
+            $template = $this->templateService->init($templateId)->getTemplate()->getBody();
+        } catch (\Exception $e) {
+            $template = "Error in body template: " . $e->getMessage();
+        }
+
+        $rows = '<tr>
```
Comment (dameert) [inline Service/Form/Submission/FormSubmissionService.php:152]: html should be generated using twig files

**Review by dameert — CHANGES_REQUESTED** [review]

File: Resources/views/email/submissions.email.twig [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+                    </td>
+                </tr>
+                <tr>
+                    <td class="email-intro"
+                        style="color: #3a383a; font-family: Georgia, serif; font-size: 18px; line-height: 27px; padding-top: 5px;">
+                        <table border="1">
+                            <tr>
+                                <th>Label</th>
+                                <th>Submission Date</th>
+                                <th>Expiry Date</th>
```
Comment (dameert) [inline Resources/views/email/submissions.email.twig:86]: With this information we can not 100% map a consentment to a user if he would ask for it.
The uuid of the submission is more important than the label. I think we should have both here.

File: Resources/views/email/submissions.email.twig [hunk 4]
```diff
@@ -0,0 +1,111 @@
...
+                            {% for submission in submissions %}
+                                <tr>
+                                    <td>{{ submission.label }}</td>
+                                    <td>{{ submission.created|date('d/m/Y') }}</td>
+                                    <td>{{ submission.expireDate|date('d/m/Y') }}</td>
+                                    {% for key, value in submission.data %}
+                                        {% if value  is not iterable %}
+                                            <td>{{ value }}</td>
+                                        {% endif %}
+                                    {% endfor %}
```
Comment (dameert) [inline Resources/views/email/submissions.email.twig:100]: What value would that be?

File: Resources/views/email/submissions.email.twig [hunk 5]
```diff
@@ -0,0 +1,111 @@
...
+                    <td class="email-intro"
+                        style="color: #3a383a; font-family: Georgia, serif; font-size: 18px; line-height: 27px; padding-top: 5px;">
+                        <table border="1">
+                            <tr>
+                                <th>Label</th>
+                                <th>Submission Date</th>
+                                <th>Expiry Date</th>
+                                {% for key, value in submissions.0.data %}
+                                    <th> {{ key }}</th>
+                                {% endfor %}
```
Comment (dameert) [inline Resources/views/email/submissions.email.twig:89]: What are you exposing this way?

**Review by dameert — CHANGES_REQUESTED** [review]

File: Resources/views/email/submissions.email.twig (hunk 5 above)
Comment (theus77) [reply Resources/views/email/submissions.email.twig:89]: All data in the submission I'm afraid. And there is a bug if submissions does not have the same fields in the same order. I understand that we need to get some information in order to know who gave its approval. Maybe we should parsed all fields of all submissions to get a list of form fields having one of those patterns in their name:['name', 'phone', 'email']?

File: Resources/views/email/submissions.email.twig (hunk 4 above)
Comment (laurensmertens) [reply Resources/views/email/submissions.email.twig:100]: The title of the column, for example "firstname" or "email", but as discussed will be removed.

File: Resources/views/email/submissions.email.twig (hunk 5 above)
Comment (laurensmertens) [reply Resources/views/email/submissions.email.twig]: Good point, at first I had it per form, but then it was indeed an issue.
Shouldn't be a problem anymore if ID, label & dates are enough.

**Review by theus77 — APPROVED** [review]

File: Command/EmailSubmissionsCommand.php [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+                null,
+                InputOption::VALUE_OPTIONAL,
+                'From which form instance do you want to mail submissions? Defaults to all.'
+            );
+    }
+
+    protected function execute(InputInterface $input, OutputInterface $output): int
+    {
+        $emails = (array) $input->getArgument('emails');
+        $formInstance = \strval($input->getOption('formInstance'));
```
Comment (dameert) [inline Command/EmailSubmissionsCommand.php:57]:
```suggestion
        $formInstance = (string) $input->getOption('formInstance');
```

File: Repository/FormSubmissionRepository.php [hunk 7]
```diff
@@ -35,6 +35,35 @@ public function findAllUnprocessed(): array
...
+    public function findAllFormSubmissions(): array
+    {
+        $qb = $this->createQueryBuilder('fs');
+        $qb
+            ->andWhere($qb->expr()->isNotNull('fs.data'))
+            ->orderBy('fs.created', 'desc');
+
+        return $qb->getQuery()->getArrayResult();
+    }
+
```
Comment (dameert) [inline Repository/FormSubmissionRepository.php:66]: findFormSubmissions(?string $formInstance = null)

File: Service/Form/Submission/FormSubmissionService.php [hunk 8]
```diff
@@ -91,4 +123,23 @@ public function submit(FormSubmissionRequest $submitRequest): array
 
         return ['submission_id' => $formSubmission->getId()];
     }
+
+    /**
+     * @param array<FormSubmission> $submissions
+     * @return string
```
Comment (dameert) [inline Service/Form/Submission/FormSubmissionService.php:129]:
```suggestion
```

File: Service/Form/Submission/FormSubmissionService.php [hunk 9]
```diff
@@ -91,4 +123,23 @@ public function submit(FormSubmissionRequest $submitRequest): array
...
+            return 'There are no submissions for this form';
+        }
+
+        try {
+            $template = $this->twig->render('@EMSCore/email/submissions.email.twig', ['submissions' => $submissions]);
+        } catch (\Exception $e) {
+            $template = $this->twig->createTemplate("Error in body template: " . $e->getMessage())->render();
+        }
+
+        return $template;
```
Comment (dameert) [inline Service/Form/Submission/FormSubmissionService.php:143]:
```suggestion
        try {
           return $this->twig->render('@EMSCore/email/submissions.email.twig', ['submissions' => $submissions]);
        } catch (\Exception $e) {
            return $this->twig->createTemplate("Error in body template: " . $e->getMessage())->render();
        }
```

File: Service/MailerService.php [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+
+namespace EMS\CoreBundle\Service;
+
+use Swift_Mailer;
+use Swift_Message;
+
+class MailerService
+{
+    const EMAIL_FROM = 'reporting@elasticms.test';
+    const NAME_FROM = 'ElasticMS';
```
Comment (dameert) [inline Service/MailerService.php:11]: Can't be hard coded. Load this from the configuration (we already have email information in the .env)

**Review by dameert — CHANGES_REQUESTED** [review]

File: Service/MailerService.php (hunk 10 above)
Comment (dameert) [reply Service/MailerService.php:11]: see:
EMS_FROM_EMAIL_ADDRESS='ems@test'
EMS_FROM_EMAIL_NAME='EMS'

File: Command/EmailSubmissionsCommand.php (hunk 6 above)
Comment (laurensmertens) [reply Command/EmailSubmissionsCommand.php]: PHPStan does not allow this - Cannot cast array<string>|bool|string|null to string.

**Review by Davidmattei — APPROVED** [review]

**Pull request merged**